analytical_engine/apps/simple_path/all_simple_paths.h

```diff
@@ -0,0 +1,324 @@
+  static constexpr grape::LoadStrategy load_strategy =
+      grape::LoadStrategy::kBothOutIn;
+  using vertex_t = typename fragment_t::vertex_t;
+  using vid_t = typename fragment_t::vid_t;
+
+  void PEval(const fragment_t& frag, context_t& ctx,
+             message_manager_t& messages) {
+    vertex_t source;
+    vid_t source_gid;
+    bool empty_flag = false;
+    // Source does not exist.
+    if (!frag.Oid2Gid(ctx.source_id, source_gid)) {
+      empty_flag = true;
+      ctx.nodes_not_found = true;
+    }
+    // nodes_not_found = true, return 2 as NodeNotFound.
+    if (ctx.nodes_not_found == true && frag.fid() == 0) {
+      std::vector<typename fragment_t::oid_t> data;
+      std::vector<size_t> shape{1};
+      data.push_back(2);
+      ctx.assign(data, shape);
+      return;
+    }
+    ctx.soucre_fid = source_gid >> ctx.fid_offset;
+    if (ctx.targets.count(source_gid) || ctx.cutoff < 1) {
+      empty_flag = true;
+    }
+    // empty_flag == true, return 1 as empty path.
+    if (empty_flag == true && frag.fid() == ctx.soucre_fid) {
+      std::vector<typename fragment_t::oid_t> data;
+      std::vector<size_t> shape{1};
+      data.push_back(1);
+      ctx.assign(data, shape);
+      return;
+    } else if (empty_flag == true || ctx.nodes_not_found == true) {
+      return;
+    }
+    bool native_source = frag.GetInnerVertex(ctx.source_id, source);
+    // If native_source resize edge_map, else send frag innervertices num to
+    // source frag.
+    if (native_source) {
+      ctx.source_flag = true;
+      size_t total_vertex_num = frag.GetTotalVerticesNum();
+      ctx.edge_map.resize(total_vertex_num);
+    } else {
+      vid_t in_num = frag.GetInnerVerticesNum();
+      fid_t fid = frag.fid();
+      std::tuple<std::pair<vid_t, vid_t>, bool, bool> msg =
+          std::make_tuple(std::make_pair((vid_t) fid, in_num), false, true);
+      messages.SendToFragment(ctx.soucre_fid, msg);
+    }
+    messages.ForceContinue();
+  }
+
+  void IncEval(const fragment_t& frag, context_t& ctx,
+               message_manager_t& messages) {
+    int frag_finish_counter = 0;
+    int init_counter = 0;
+    std::tuple<std::pair<vid_t, vid_t>, bool, bool> msg;
+
+    while (messages.GetMessage(msg)) {
+      vid_t gid = std::get<0>(msg).first;
+      vid_t msg2 = std::get<0>(msg).second;
+      // msg[2] == true means init frag_vertex_num, msg[1] == false means new
+      // vertex to deal with, msg[1] == true means update edge_map.
+      if (std::get<2>(msg) == true) {
+        init_counter++;
+        ctx.frag_vertex_num[gid] = msg2;
+        if (init_counter == static_cast<int>(frag.fnum()) - 1) {
+          vertex_t source;
+          frag.GetInnerVertex(ctx.source_id, source);
+          vertexProcess(source, frag, ctx, messages, 0);
+          vid_t s_gid;
+          frag.Oid2Gid(ctx.source_id, s_gid);
+          ctx.visit.insert(s_gid);
+          messages.ForceContinue();
+          frag_finish_counter++;
+        }
```

Comment:
these code should not as part of message processing, you should add a `step` in ctx and use the `step` to take care of these initialization.  
```
if (ctx.step == 1) {
    process the frag vnum messages
    vertexProcessing
} else {
   take care other type messages
}
````